### PR TITLE
fix a bug in CachedOp.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
After adding kSubgraphExec, some logic in CachedOp is no longer valid. For example, an op executor that requires async execution may not contain output arrays. On the other hand, calling `CreateEngineOp` on op executors without output arrays can fail. This PR tries to fix this problem.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: tests/python/unittest/test_gluon_rnn.py [hunk 1]
```diff
@@ -64,25 +64,31 @@ def check_contrib_rnn(cell_type, num_states):
...
-
-    trainer = gluon.Trainer(params2, 'sgd', {'learning_rate' : 0.03})
-    with mx.autograd.record():
+    configs = [
+            {},
+            {'static_alloc': True},
+            {'static_alloc': True, 'static_shape': True} ]
+    for config in configs:
+        layer = TestRNNLayer(cell_type, hidden_size)
+        layer.initialize(ctx=mx.cpu(0))
```
Comment (marcoabreu) [inline tests/python/unittest/test_gluon_rnn.py]: Are we sure about hardcoding the context?

File: src/imperative/imperative_utils.h [hunk 2]
```diff
@@ -963,13 +963,13 @@ inline void CreateEngineOpSeg(
...
+      seg = EngineOprSeg{false, nid + 1, nullptr};
       seg_execs.clear();
       seg_start = nid + 1;
-    } else if (!valid) {
-      seg = EngineOprSeg{false, nid + 1, nullptr};
+    } else if (is_async) {
+      seg = EngineOprSeg{false, nid + 1};
+      seg.opr.reset(CreateEngineOp(default_ctx, seg_execs));
       seg_execs.clear();
       seg_start = nid + 1;
```
Comment (szha) [inline src/imperative/imperative_utils.h]: extract common code outside branch

**Comment by zheng-da** [comment]
the PR will be updated after https://github.com/apache/incubator-mxnet/pull/11566 is merged.

File: src/imperative/imperative_utils.h (hunk 2 above)
Comment (zheng-da) [reply src/imperative/imperative_utils.h]: we can't. the code is used in "if" and "else if". it's not used in "else"

**Comment by eric-haibin-lin** [comment]
what's the new test case added for this?

**Comment by zheng-da** [comment]
It's here. I tested the code in more CachedOp configurations.
https://github.com/apache/incubator-mxnet/pull/11675/files#diff-52501b7b512a5434dbc54931da1f0f2cR1006
